package NReqBundleIteratorProto;

message TPositionArray {
    // Raw binary representation, see NReqBundleIterator::TPosition::Raw
    repeated uint64 RawPositions = 1 [packed=true];
}

// See NDoom::TRangedKeyId
message TOffroadRangedKeyId {
    optional uint32 Id = 1;
    optional uint32 Range = 2;
}

message TOffroadIterator {
    optional TOffroadRangedKeyId OffroadTermId = 1;
    optional TPositionArray Templates = 2;
}

message TAndIterator {
    repeated TBlockIterator Iterators = 1;
    optional fixed64 FullHitsMask = 2;
    optional uint32 MaxDistance = 3;
    optional uint32 NumWords = 4;
}

message TOrderedAndIterator {
    repeated TBlockIterator Iterators = 1;
    optional fixed64 AnyWordsMask = 2;
    optional uint32 NumWords = 4;
}

// Only for ItIndex iterators
message TIndexIterator {
    // Numerical value of NReqBundleIteratorImpl::EIteratorType (one of DefaultIteratorType, StopWordIteratorType, AttributeIteratorType)
    optional uint32 IteratorType = 1;
    oneof Variant {
        TOffroadIterator OffroadIterator = 2;

        // Yndex iterators are not supported (yet?)
    }
}

message TBlockIterator {
    oneof Variant {
        TIndexIterator IndexIterator = 1;
        TAndIterator AndIterator = 2;
        TOrderedAndIterator OrderedAndIterator = 3;
    }
    optional uint32 BlockId = 4;
    optional uint32 BlockType = 5;
}

message TRbIterator {
    repeated TBlockIterator Iterators = 1;
    repeated uint32 RichTreeFormIds = 2 [packed=true];
    repeated uint32 RichTreeBlockOffsets = 3;
}
